[mono][NativeAOT] System.Globalization.Native build improvements

filipnavara opened the pull request:
Fixes #82389

- Fixes build of System.Globalization.Native for iOS-like, browser, and wasi platforms
- Removes ICU shim from static builds of Mono runtime and builds it as part of libs.native subset
- Switches browser/wasi to use the static libraries

filipnavara committed:
Add System.Globalization.Native builds for iOS-like platforms

filipnavara commented on src/native/libs/System.Globalization.Native/CMakeLists.txt:
```diff
@@ -69,6 +69,10 @@ if (LOCAL_BUILD)
     include_directories(${CMAKE_CURRENT_BINARY_DIR})
+elseif (CLR_CMAKE_TARGET_MACCATALYST OR CLR_CMAKE_TARGET_IOS OR CLR_CMAKE_TARGET_TVOS)
+    add_definitions(-DSTATIC_ICU)
+    add_definitions(-DPALEXPORT=EXTERN_C)
```
```suggestion
```

filipnavara commented on src/native/libs/System.Globalization.Native/pal_collation.c:
```diff
@@ -335,7 +335,10 @@ static UCollator* CloneCollatorWithOptions(const UCollator* pCollator, int32_t o
 #if HAVE_UCOL_CLONE
         pClonedCollator = ucol_clone(pCollator, pErr);
 #else
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
         pClonedCollator = ucol_safeClone(pCollator, NULL, NULL, pErr);
+#pragma clang diagnostic pop
```
The compilation uses Homebrew ICU headers which are newer (>=71) than the shipped ICU.

filipnavara commented:
The build failures are relevant. I'll try to remove the duplicate build from Mono sources.

filipnavara commented:
This fundamentally broken at the moment. We would need to link to the packages obtained from Microsoft.NETCore.Runtime.ICU.Transport. I'll have to rethink how to do that.

filipnavara committed:
Update package manifest

filipnavara committed:
Move copying of ICU static libs and data files from Mono to libs.native

filipnavara committed:
Mono: Link static ICU shim only into Mono shared libraries

filipnavara committed:
Include ICU .dat files in LibrariesRuntimeFiles

filipnavara committed:
Build System.Globalization.Native on Browser/WASI

filipnavara committed:
Fix configuration options for WASM/WASI builds

filipnavara committed:
Update src/libraries/native-binplace.proj to include ICU .dat files

filipnavara committed:
List correct libraries in link_libraries

filipnavara committed:
Fix iOS build after the link fix :/

filipnavara committed:
Use correct ICU build for browser-wasm-threads builds

filipnavara committed:
Attempt to fix WASI build

filipnavara committed:
Fix libSystem.Globalization.Native.a output name for WASI

filipnavara committed:
Fix merge

filipnavara committed:
Fix multithread browser builds

filipnavara committed:
Fix WASI linking

filipnavara commented:
The WASI part has overlap with #82444.

am11 commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -63,8 +63,7 @@ The .NET Foundation licenses this file to you under the MIT license.
       <NetCoreAppNativeLibrary Include="System.Native" />
-      <!-- FIXME: The library is currently not available for iOS-like platforms -->
-      <NetCoreAppNativeLibrary Include="System.Globalization.Native" Condition="'$(StaticICULinking)' != 'true' and '$(_IsiOSLikePlatform)' != 'true'" />
+      <NetCoreAppNativeLibrary Include="System.Globalization.Native" Condition="'$(StaticICULinking)' != 'true'" />
```
@filipnavara, in case it wasn't obvious: the `StaticICULinking` is a bit different. It is meant to run cmake _during_ `dotnet publish -p:PublishAot=true` on user machine. I'm not sure how the variant added in `src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props` will be used here, but we seem to have two other options:

1. either bring iOS to the same plan as existing platforms (run cmake to build Globalization shim during dotnet-publish on user machine).
2. or use it from the framework directory; change line 89 as:
    ```diff
    -       <NativeLibrary Include="$(IntermediateOutputPath)libs/System.Globalization.Native/build/libSystem.Globalization.Native.a" />
    +       <NativeLibrary Include="$(IntermediateOutputPath)libs/System.Globalization.Native/build/libSystem.Globalization.Native.a" Condition="'$(_IsiOSLikePlatform)' != 'true'" />
    +       <NativeLibrary Include="$(IlcFrameworkNativePath)libSystem.Globalization.Native.a" Condition="'$(_IsiOSLikePlatform)' == 'true'" />
    ```
      * Note that we already have (an unused!) copy of `libSystem.Globalization.Native.a` in the framework directory:
          ```zsh
          % ls ~/.nuget/packages/runtime.linux-arm64.microsoft.dotnet.ilcompiler/8.0.0-preview.2.23116.1/framework/*Globalization*.a
          /Users/am11/.nuget/packages/runtime.linux-arm64.microsoft.dotnet.ilcompiler/8.0.0-preview.2.23116.1/framework/libSystem.Globalization.Native.a
          ```
          which comes from https://github.com/dotnet/runtime/blob/af18fd07d1bb5a0d923a49c39b19c6fbcb6dc07b/eng/liveBuilds.targets#L175 (unused files are  missing from the ilc package excludes, which should be cleaned up separately)

I'm not familiar with NativeAOT step in Xamrian publish workflow, but if option 1 is feasible for iOS-like platforms, I'd go for it.

filipnavara replied:
libSystem.Globalization.Native.a was not built for iOS-like platforms, so that's part of the fix.

The final linking is done either through AppleAppBuiler or the Xamarin tooling. Both of them can consume the .a file once it exists.

`StaticICULinking` remains unused for the moment for these platforms.

am11 replied:
> `StaticICULinking` remains unused for the moment for these platforms.

I missed that! 😁

> libSystem.Globalization.Native.a was not built for iOS-like platforms, so that's part of the fix.

Cool. For NativeAOT, does Xamarin use shared libraries from `microsoft.dotnet.ilcompiler` package or from the Microsoft.NETCore.App / shared framework? e.g. `libSystem.Globalization.Native.a` _can_ be different if user has csproj file with:

```csproj
<!-- Different version than what comes with SDK https://github.com/dotnet/installer/blob/8ce7bb903a8d818bd288267f82be0d98f5e6c757/src/redist/targets/GenerateBundledVersions.targets#L408 -->
<PackageReference Include="Microsoft.DotNet.ILCompiler" Version="$(DifferentVersion)" />
```

filipnavara replied:
That's kind of unanswered packaging question at the moment. 😅

For the current MonoVM-based workloads it comes from Microsoft.NETCoreApp.Runtime.Mono.<RID> packages. This PR adds libSystem.Globalization.Native.a into the package and removes the same APIs from libmonosgen.a (or whether is the static runtime name).

ivanpovazan replied:
> StaticICULinking remains unused for the moment for these platforms.
Should we add a MSBuild warning/error when someone tries to use it on these platforms?

filipnavara replied:
> Should we add a MSBuild warning/error when someone tries to use it on these platforms?

It would be nice to have explicit check but I would prefer not to mix it into this PR.

I don't touch any of the `StaticICULinking` code here and it's documented to work only on Linux when not cross-compiling:
https://github.com/dotnet/runtime/blob/cabaab5e90e19256ae404970091350bb6f0f7139/src/coreclr/nativeaot/docs/compiling.md?plain=1#L77-L85

filipnavara committed:
Revert WASI changes after rebase

filipnavara committed:
Try to fix builds on Windows

pavelsavara commented on src/native/libs/Common/pal_atomic.h:
```diff
@@ -4,7 +4,7 @@
 #pragma once
 
-#if defined(TARGET_UNIX)
+#if defined(TARGET_UNIX) || defined(TARGET_WASI)
```
Browser too, for consistency ?

filipnavara replied:
I think the browser builds currently define `TARGET_UNIX`. I'd expect WASI to do the same but I opted for minimal changes.

pavelsavara replied:
Browser has very good posix emulation, so you are right about `TARGET_UNIX`. I would still prefer more explicit reference.
WASI is not unix.

filipnavara replied:
While I understand your reasoning I hesitate to do the change to reduce collateral damage.

Prior to this PR the ICU shim was built in the Mono runtime and it set `TARGET_UNIX` for both for browser ([here](https://github.com/dotnet/runtime/blob/cabaab5e90e19256ae404970091350bb6f0f7139/src/mono/CMakeLists.txt#L696)) and WASI ([here](https://github.com/dotnet/runtime/blob/cabaab5e90e19256ae404970091350bb6f0f7139/src/mono/CMakeLists.txt#L691)). 

In the new setup the System.Globalization.Native library shares the same build system as System.Native. The existing System.Native build defines TARGET_UNIX for Browser, but not for WASI.

If I add `|| defined(TARGET_BROWSER)` here it will be strictly redundant (since `TARGET_BROWSER` implies `TARGET_UNIX`) and inconsistent with the rest of the System.Native code that has no such checks.

akoeplinger replied:
I'm OK with this difference here for now, though we should probably move away from setting TARGET_UNIX for browser too. For example in the managed src/libraries build we do _not_ set `TargetsUnix` msbuild property which gets used to set `TARGET_UNIX` C# define for neither Browser nor Wasi.

pavelsavara commented on src/native/libs/System.Globalization.Native/CMakeLists.txt:
```diff
@@ -1,19 +1,23 @@
 project(System.Globalization.Native C)
 
-if(CLR_CMAKE_TARGET_UNIX)
+if(CLR_CMAKE_TARGET_UNIX OR CLR_CMAKE_TARGET_WASI)
```
Browser ?

pavelsavara commented on src/native/libs/System.Globalization.Native/pal_calendarData.c:
```diff
@@ -9,7 +9,7 @@
 #include "pal_calendarData.h"
 
-#if defined(TARGET_UNIX)
+#if defined(TARGET_UNIX) || defined(TARGET_WASI)
```
Browser

pavelsavara commented on src/libraries/native-binplace.proj:
```diff
@@ -24,6 +24,7 @@
       <BinPlaceItem Include="$(NativeBinDir)*.dwarf" />
       <BinPlaceItem Include="$(NativeBinDir)*.dex" />
       <BinPlaceItem Include="$(NativeBinDir)*.jar" />
+      <BinPlaceItem Include="$(NativeBinDir)*.dat" />
```
is this dealing also with ICU shards ?

pavelsavara replied:
For browser we have this https://github.com/dotnet/runtime/pull/80421 
and for WASI we still need more https://github.com/dotnet/runtime/issues/82567

filipnavara replied:
It copies all the icu*.dat files just like mono.proj did.

filipnavara replied:
On second though, it possibly makes `ICULibNativeFiles` and `ICULibFiles` copying in wasm.proj and wasi.proj redundant.

pavelsavara commented on src/mono/wasm/runtime/CMakeLists.txt:
```diff
@@ -27,6 +27,7 @@ target_link_libraries(dotnet
     ${MONO_ARTIFACTS_DIR}/libmono-profiler-aot.a
     ${MONO_ARTIFACTS_DIR}/libmono-profiler-browser.a
     ${NATIVE_BIN_DIR}/libSystem.Native.a
+    ${NATIVE_BIN_DIR}/libSystem.Globalization.Native.a
```
I must admit I don't understand big picture. Could you please explain the consequences of this for browser ?

filipnavara replied:
Instead of building the ICU shim as part of the Mono runtime build, it moves the build to libs.native part of the build. Then it links the resulting static library back.

The end result should be roughly identical to the previous status quo.

There are several reasons for this change:
- Reuse of the build system in libs.native instead of maintaining a second copy in Mono build
- iOS-like platforms need the library for both Mono and NativeAOT, so building it as part of Mono runtime is problematic; it can be shared by both runtimes this way
- Browser/WASI are just changed to follow the same pattern to make the build system less complicated

pavelsavara replied:
thanks!

lambdageek replied:
/cc @steveisok LibraryBuilder targets should also be aware of this change (although I suspect it doesn't matter to the library builder where the static lib came from, as long as it's there)

akoeplinger replied:
from what I can see the LibraryBuilder doesn't need updating since it just takes all *.a files

akoeplinger commented on src/native/libs/System.Globalization.Native/CMakeLists.txt:
```diff
@@ -1,19 +1,23 @@
 
-    if (NOT CLR_CMAKE_TARGET_ANDROID)
+    # pal_icushim_static.c(41,24): error G026380BD: unused function 'icu_trace_data'
+    add_compile_options(-Wno-unused-function)
```
we should instead wrap the `icu_trace_data` function with `#if defined(ICU_TRACING)`

pavelsavara commented on src/native/libs/System.Globalization.Native/CMakeLists.txt:
```diff
@@ -59,6 +63,18 @@ set(NATIVEGLOBALIZATION_SOURCES
+if (DEFINED CMAKE_ICU_DIR)
+    include_directories(${CMAKE_ICU_DIR}/include)
+    link_libraries(${CMAKE_ICU_DIR}/lib/libicuuc.a ${CMAKE_ICU_DIR}/lib/libicui18n.a ${CMAKE_ICU_DIR}/lib/libicudata.a)
+    link_libraries(stdc++)
```
Also https://github.com/dotnet/runtime/pull/82444/files#diff-fbc587d0b04feb58ec4c10941c3beb54321a1c10d0ffb09c44ebee0100a58103R21

filipnavara replied:
It always happens somewhere, either at linking the runtime as dynamic library, or as part of the final linking of the application.

In this particular case we happen to build .dylib for iOS which may not be strictly necessary. Since the project is marked as using the "C" language and not "CXX" the C++ standard library is not included implicitly.

We can exclude the .dylib builds as an alternative.

akoeplinger replied:
Ok, I'm mostly worried about Android where we can't introduce this stdc++ dependency. For Android we're loading ICU .so via dlopen from the system location so we just need to make sure this doesn't apply for Android.

I'd probably move the two `link_libraries` down to the condition where we set STATIC_ICU for iOS/Catalyst etc

filipnavara replied:
On Android you will end up linking to the libmono*.so that has the ICU shim built-in. This PR doesn't affect that. We do build an additional libSystem.Globalization.Native.[a/so] and it likely is not necessary. While not necessary for the supported configurations, it is nice to have for the unsupported NativeAOT/Android builds.

filipnavara commented on src/native/libs/System.Globalization.Native/CMakeLists.txt:
```diff
@@ -1,19 +1,23 @@
     add_compile_options(-Wno-reserved-id-macro)
     add_compile_options(-Wno-documentation)
     add_compile_options(-Wno-documentation-unknown-command)
+    add_compile_options(-Wno-reserved-identifier)
```
There's some reserved identifier in the ICU headers (prefixed with `__`). Newer emsdk compiler triggers slightly different warning than the ones that were already suppressed for the same reason.

akoeplinger commented on src/native/libs/System.Globalization.Native/CMakeLists.txt:
```diff
@@ -97,6 +112,9 @@ if (GEN_SHARED_LIB)
         find_library(FOUNDATION Foundation REQUIRED)
     endif()
+    if (CLR_CMAKE_TARGET_MACCATALYST OR CLR_CMAKE_TARGET_IOS OR CLR_CMAKE_TARGET_TVOS)
+        find_library(ICUCORE icucore REQUIRED)
```
not sure where this will actually find icucore from. I think I'd prefer turning off GEN_SHARED_LIB for iOS/Catalyst etc in general for System.Globalization.Native, it wouldn't do the right thing in the non-static case.

akoeplinger replied:
or we need to make a .dylib of the "static" version where it links in the icu .a

filipnavara replied:
This is actually a leftover that I forgot to remove. The branch is never taken on the iOS-like platforms.

filipnavara committed:
Remove unnecessary icucore change

filipnavara committed:
Remove unnecessary empty line

filipnavara committed:
Wrap icu_trace_data in `#if defined(ICU_TRACING)`

kotlarmilos commented:
@filipnavara is there anything for this PR or is it waiting for a review?

filipnavara commented:
> @filipnavara is there anything for this PR or is it waiting for a review?

Waiting for review.

ivanpovazan reviewed (approved):
Minor note: The title seems outdated as the PR changes now affect mono builds and other platforms as well. I don't have a strong opinion but something like `[mono][NativeAOT] System.Globalization.Native build improvements` might be more suitable.

Other than that, LGTM.
Thank you very much!

PS I would leave this to @akoeplinger for final approval.

akoeplinger committed:
Merge branch 'main' into icu-ios

akoeplinger committed:
Fix "unused parameter 'symbolVersion'" error

akoeplinger committed:
Fix NativeAOT build for Apple Mobile

The coreclr CMake "includes" the src/native directory so it doesn't go through build-native.sh.

akoeplinger commented:
Failures are unrelated.

filipnavara commented:
> did anyone check the final post linking binary sizes on this?

Any specific configuration you are concerned about?

For NativeAOT/iOS there's nothing to compare against since it didn't work previously. For everything else it shuffles a bit where/how the globalization shim is built but it doesn't really add or remove any code, or at least not intentionally.

steveisok commented:
> did anyone check the final post linking binary sizes on this?

Quick spot check on one of Alex's test official builds show no change.  I'll check again when a real one gets through.
